src/api/classify_element.rs

```diff
@@ -0,0 +1,349 @@
+}
+
+impl<E: ClassifyElement> ClassifyElementProvider<E> {
+    fn new(
+        from_consumer: crossbeam_channel::Receiver<Option<E::Packet>>,
+        await_consumer: crossbeam_channel::Sender<task::Task>,
+        wake_consumer: crossbeam_channel::Receiver<task::Task>
+    ) -> Self {
+        ClassifyElementProvider {
+            from_consumer,
+            await_consumer,
+            wake_consumer
+        }
+    }
+}
+
+impl<E: ClassifyElement> Drop for ClassifyElementProvider<E> {
+    fn drop(&mut self) {
+        if let Ok(task) = self.wake_consumer.try_recv() {
+            task.notify();
+        }
+    }
+}
+
+impl<E: ClassifyElement> Stream for ClassifyElementProvider<E> {
+    type Item = E::Packet;
+    type Error = ();
+
+    ///Implement Poll for Stream for AsyncElementProvider
+    /// 
+    /// This function, tries to retrieve a packet off the `from_consumer`
+    /// channel, there are four cases: 
+    /// ###
+    /// #1 Ok(Some(Packet)): Got a packet. If the consumer needs, (likely due to 
+    /// an until-now full channel) to be awoken, wake them. Return the Async::Ready(Option(Packet))
+    /// 
+    /// #2 Ok(None): this means that the consumer is in tear-down, and we
+    /// will no longer be receivig packets. Return Async::Ready(None) to forward propagate teardown
+    /// 
+    /// #3 Err(TryRecvError::Empty): Packet queue is empty, await the consumer to awaken us with more
+    /// work, and return Async::NotReady to signal to runtime to sleep this task.
+    /// 
+    /// #4 Err(TryRecvError::Disconnected): Consumer is in teardown and has dropped its side of the
+    /// from_consumer channel; we will no longer receive packets. Return Async::Ready(None) to forward
+    /// propagate teardown.
+    /// ###
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.from_consumer.try_recv() {
+            Ok(Some(packet)) => {
+                if let Ok(task) = self.wake_consumer.try_recv() {
+                        task.notify();
+                }
+                Ok(Async::Ready(Some(packet)))
+            },
+            Ok(None) => {
+                Ok(Async::Ready(None))
+            },
+            Err(TryRecvError::Empty) => {
+                let task = task::current();
+                if let Err(_) = self.await_consumer.try_send(task) {
+                    task::current().notify();
+                }
+                Ok(Async::NotReady)
+            },
+            Err(TryRecvError::Disconnected) => {
+                Ok(Async::Ready(None))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::api::element::{Element};
+    use crate::utils::test::packet_generators::{immediate_stream, PacketIntervalGenerator};
+    use crate::utils::test::packet_collectors::ExhaustiveCollector;
+    use crossbeam::crossbeam_channel;
+    use core::time;
+
+    use futures::future::lazy;
+
+    #[allow(dead_code)]
+    struct IdentityElement {
```

Comment:
IdentityElement should definitely be defined elsewhere for testing purposes so we don't have to keep redefining it. We can do that in a separate commit.

Reply:
Yeah I can open an issue

Reply:
https://github.com/route-rs/route-rs/issues/30